corehq/apps/userreports/expressions/specs.py

```diff
@@ -263,3 +266,39 @@ def get_variables(self, item, context):
             for slug, variable_expression in self._context_variables.items()
         }
         return var_dict
+
+
+class FormsExpressionSpec(JsonObject):
+    type = TypeProperty('get_case_forms')
+    case_id_expression = DefaultProperty(required=True)
+
+    def configure(self, case_id_expression):
+        self._case_id_expression = case_id_expression
+
+    def __call__(self, item, context=None):
+        case_id = self._case_id_expression(item, context)
+
+        if not case_id:
+            return []
+
+        assert context.root_doc['domain']
+        return self._get_forms(case_id, context)
+
+    def _get_forms(self, case_id, context):
+        domain = context.root_doc['domain']
+
+        cache_key = (self.__class__.__name__, case_id, domain)
+
+        if context.get_value(cache_key) is not None:
```

Comment:
as a basic pass it looks good. 

It's a bit nitpicky/subjective, but I might favor just exposing `.cache` and getting rid of `get_value`/`set_value` just to keep things really explicit. or if you like keeping the abstraction maybe just changing the functions to `get_cached_value`/`set_cached_value` so it's obvious what that's doing behind the scenes.